File: November 2024 Leetcode Solution/November-27.java
import java.util.*;
class Solution {
    public int[] shortestDistanceAfterQueries(int n, int[][] queries) {
        List<Set<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new HashSet<>());
        }
        for (int i = 0; i < n - 1; i++) {
            graph.get(i).add(i + 1); 
        }
        int[] result = new int[queries.length];
        int[] dist = new int[n]; 
        for (int i = 0; i < queries.length; i++) {
            int u = queries[i][0], v = queries[i][1];
            if (!graph.get(u).contains(v)) {
                graph.get(u).add(v); 
            }
            Arrays.fill(dist, -1); 
            result[i] = bfs(graph, dist, n);
        }
        return result;
    }
    private int bfs(List<Set<Integer>> graph, int[] dist, int n) {
        dist[0] = 0;
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(0);
        while (!queue.isEmpty()) {
            int u = queue.poll();
            for (int v : graph.get(u)) {
                if (dist[v] == -1) {
                    dist[v] = dist[u] + 1;
                    if (v == n - 1) {
                        return dist[v];
                    }
                    queue.offer(v);
                }
            }
        }
        return -1; 
    }
}

File: November 2024 Leetcode Solution/November-27.c
int* shortestDistanceAfterQueries(int n, int** queries, int queriesSize, int* queriesColSize, int* returnSize) {
    *returnSize = queriesSize;
    int* result = (int*)malloc(queriesSize * sizeof(int));
    int** graph = (int**)malloc(n * sizeof(int*));
    int* sizes = (int*)calloc(n, sizeof(int));
    
    for (int i = 0; i < n; ++i) {
        graph[i] = (int*)malloc(n * sizeof(int));
        if (i < n - 1) graph[i][sizes[i]++] = i + 1;
    }
    
    int* dist = (int*)malloc(n * sizeof(int));
    
    void bfs() {
        memset(dist, -1, n * sizeof(int));
        dist[0] = 0;
        int* queue = (int*)malloc(n * sizeof(int));
        int front = 0, back = 0;
        queue[back++] = 0;
        
        while (front < back) {
            int u = queue[front++];
            for (int i = 0; i < sizes[u]; ++i) {
                int v = graph[u][i];
                if (dist[v] == -1) {
                    dist[v] = dist[u] + 1;
                    queue[back++] = v;
                }
            }
        }
        free(queue);
    }
    
    bfs();
    
    for (int i = 0; i < queriesSize; ++i) {
        int u = queries[i][0], v = queries[i][1];
        graph[u][sizes[u]++] = v;
        if (dist[v] == -1 || dist[v] > dist[u] + 1) {
            bfs();
        }
        result[i] = dist[n - 1];
    }
    
    for (int i = 0; i < n; ++i) free(graph[i]);
    free(graph);
    free(sizes);
    free(dist);
    return result;
}
